lib/tbd/ua.rb

```diff
@@ -183,151 +187,163 @@ def uprate(model = nil, s = {}, argh = {})
 
       if g[:op].empty?
-        log(ERR, "Construction to uprate? (#{mth})")
+        log(ERR, "Construction (#{type}) to uprate? (#{mth})")
       elsif all
-        model.getSurfaces.each do |sss|
-          next unless sss.surfaceType.downcase.include?(label.to_s)
-          next unless sss.outsideBoundaryCondition.downcase == "outdoors"
-          next     if sss.construction.empty?
-          next     if sss.construction.get.to_LayeredConstruction.empty?
-
-          c         = sss.construction.get.to_LayeredConstruction.get
-          i         = c.nameString
-
-          # Reliable unless referenced by other surface types e.g. floor vs wall.
-          if c.getNetArea > area
-            area = c.getNetArea
+        s.each do |nom, surface|
+          next unless surface.key?(:deratable   )
+          next unless surface.key?(:type        )
+          next unless surface.key?(:construction)
+          next unless surface.key?(:filmRSI     )
+          next unless surface.key?(:index       )
+          next unless surface.key?(:ltype       )
+          next unless surface.key?(:r           )
+          next unless surface[:deratable   ]
+          next unless surface[:type        ] == typ
+          next unless surface[:construction].is_a?(cl3)
+          next     if surface[:index       ].nil?
+
+          # Retain lowest surface film resistance (e.g. tilted surfaces).
+          c    = surface[:construction]
+          i    = c.nameString
+          aire = c.getNetArea
+          film = surface[:filmRSI] if surface[:filmRSI] < film
+
+          # Retain construction covering largest area. The following conditional
+          # is reliable UNLESS linked to other deratable surface types e.g. both
+          # floors AND walls (see "elsif lc" corrections below).
+          if aire > area
             lc   = c
+            area = aire
             id   = i
           end
 
-          film    = sss.filmResistance if sss.filmResistance < film
-          nom     = sss.nameString
-          coll[i] = { area: c.getNetArea, lc: c, s: {} } unless coll.key?(i)
-          coll[i][:s][nom] = { a: sss.netArea } unless coll[i][:s].key?(nom)
+          coll[i]          = { area: aire, lc: c, s: {} } unless coll.key?(i)
+          coll[i][:s][nom] = { a: surface[:net] }  unless coll[i][:s].key?(nom)
         end
       else
         id = g[:op]
-        c  = model.getConstructionByName(id)
-
-        if c.empty?
-          log(ERR, "Construction '#{id}'? (#{mth})")
-        else
-          c  = c.get.to_LayeredConstruction
-
-          if c.empty?
-            log(ERR, "'#{id}' layered construction? (#{mth})")
-          else
-            lc       = c.get
-            area     = lc.getNetArea
-            coll[id] = { area: area, lc: lc, s: {} }
-
-            model.getSurfaces.each do |sss|
-              next unless sss.surfaceType.downcase.include?(label.to_s)
-              next unless sss.outsideBoundaryCondition.downcase == "outdoors"
```

Comment:
2nd instance of insulated attic floors not caught by TBD (when uprating), i.e. such floors are part of the envelope, yet not facing "outdoors".